pkg/image/index/index.go

```diff
@@ -0,0 +1,179 @@
+package index
+
+import (
+	"time"
+
+	"github.com/docker/distribution/digest"
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/client"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type ImageIndex struct {
+	client client.ImageInterface
+
+	imageController  *framework.Controller
+	imageIndex       cache.Indexer
+	imageStoreSynced func() bool
+}
+
+var (
+	ChainIndexName     = "chain"
+	LastChainIndexName = "last-chain"
+	// TODO: Re-listing images might be expensive on large clusters, this should be
+	// configurable?
+	ResyncInterval = 10 * time.Minute
+)
+
+// NewImageIndex returns a new image index.
+func NewImageIndex(client client.ImageInterface, stopChan <-chan struct{}) *ImageIndex {
+	c := &ImageIndex{
+		client: client,
+	}
+	c.imageIndex, c.imageController = framework.NewIndexerInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return c.client.List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return c.client.Watch(options)
+			},
+		},
+		&imageapi.Image{},
+		ResyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    c.addImage,
+			UpdateFunc: c.updateImage,
+			DeleteFunc: c.deleteImage,
+		},
+		cache.Indexers{
+			ChainIndexName:     chainIndexFunc,
+			LastChainIndexName: lastChainIndexFunc,
+		},
+	)
+
+	go c.imageController.Run(stopChan)
+	return c
+}
+
+func (c *ImageIndex) WaitForSyncedStores() {
+	for !c.imageController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// Descendants provides a list of images that the provided image is based on.
+// Note that if a base image is tagged multiple times, this will return all the matching
+// images.
+func (c *ImageIndex) Descendants(image *imageapi.Image) ([]*imageapi.Image, error) {
```

Comment:
@miminar You can have more than one image with  the same sequence of layers. You can't determine who exactly did you follow. You can only say that these images contain the sequence of layers of my ancestor.